src/verbosity.jl

```diff
@@ -0,0 +1,89 @@
+function restore_callsite_source_position!(expr, src)
+    # used to fix the logging source file + line
+    # since we're lowering our verbose logging macros to the
+    # Logging.jl macros; otherwise, they would always report this (verbosity.jl)
+    # file as the logging callsite
+    expr.args[1].args[2] = src
+    return expr
+end
+
+vlogmacrodocs = """
+    @debugv N msg args...
+    @infov N msg args...
+    @warnv N msg args...
+    @errorv N msg args...
+
+"Verbose" logging macros. Drop in replacements of standard logging macros, but an
+additional verbosity level `N` is passed to indicate differing verbosity levels
+for a given log level. The verbosity argument is subtracted from the base log level when passed down
+to the core logging logic, so `@debugv 1 msg` will essentially call `@logmsg Debug-1 msg`.
+
+An `LoggingExtras.LevelOverrideLogger`can then be used to filter on the `level` argument.
+For convenience, the
+[`LoggintExtras.with(f; level, verbosity)`](@ref) function is provided to temporarily
+wrap the current logger with a log level and verbosity subtracted to filter while `f` is executed.
+"""
+
+macro debugv(verbosity::Int, msg, exs...)
+    return restore_callsite_source_position!(
+        esc(:($Base.@logmsg (Logging.Debug - $verbosity) $msg $(exs...))),
```

Comment:
I don't think so?

Reply:
Why not? Now this will not be within the Debug level range

Reply:
Or I guess, `+` instead of `-`.

Reply:
For me, I think of it like this: there are _really_ two dimensions we're talking about: severity and verbosity. Sometimes, people view the log level _severity_ spectrum as also corresponding to verbosity, but that's not quite right; the severity is more about the _type_ of log, and how it should be seen in various scenarios. i.e. `Debug` is for non-normal/production cases where you're trying to see additional development-related information that's going on. `Info` is general things that happen/major changes in state or milestones. `Warn` is there's something non-fatal that was encountered. `Error` is something bad happened.

Verbosity, on the other hand, is the _level of detail_ we want to see _for a given severity_. Now, I'll readily admit this principle doesn't seem super applicable/useful for non-Debug log levels, but for debugging, it really does make a strong point, and indeed many systems also include a `Trace` log level in this vein: sometimes you want _even more_ debug information than `Debug`. But I would argue even for `Error` or `Warn`, there are theoretical cases where you might want to emit _more verbose_ messages than the default if requested by the user (e.g. for an `@inbounds` error, maybe I want to see the full collection being indexed, where as by default you almost surely want some kind of compact printing).

All that to say, instead of _actually_ introducing two dimensions for severity and verbosity, we're hijacking the fact that the log levels themselves are so spread out in the integer domain to say that log levels in between the standard severity levels correspond to _verbosity_ instead of actually different categorical/types of logs. So in the same vein of the severity spectrum at least roughly/partly ordered by verbosity (i.e. I would expect more `Trace` or `Debug` messages than `Error` or `Warn`), we're saying that `Debug - 1` is _like_ `Debug`, but one level _more verbose_. And `Debug - 2` is also like debug, but 2 levels more verbose, and 1 more level verbose than `Debug - 1`. So the levels _less than_ the severity level, but _greater than_ the next lower severity level constitute the possible "verbosity" levels for that severity level.

Reply:
I just find it really strange that a verbose info message is a debug-level message, a verbose warning is a info-level message, and verbose error is a warning-level message.

Reply:
Why do you say that a verbose info is a debug level? In my mind Info is LogLevel(0), Debug is LogLevel(-1000); in between is.....not well defined? I don't see any reason LogLevel(-1) would necessarily be "debug" related more than info" related

Reply:
See https://julialogging.github.io/tutorials/logging-basics/#Log-levels

> a log message with level X is a:
>  - debug message if Logging.Debug <= X < Logging.Info,
>  - info message if Logging.Info <= X < Logging.Warn,
>  - warn message if Logging.Warn <= X < Logging.Error,
>  - error message if Logging.Error <= X < Logging.AboveMaxLevel.

So a verbose info message (`@infov 1`) gets level `Logging.Info - 1`, which falls within the debug levels.

Reply:
You can also see this from the color of the log message in the REPL.

Reply:
Huh, I didn't know/realize that! What's the reasoning there (if you know) for having the levels _above_ correspond to the level instead of below? As I explained above, o was just going off the intuition where log levels are somewhat a spectrum for verbosity (I.e. debug tends to be more verbose than info), and thus _more_ verbose than Debug would be Debug-1. 

In reality, I don't think it really matters for the verbosity stuff because we're just using the "nearby" log levels somewhat arbitrarily. But it sounds nice to get the color printing to lineup. It'd be nice to document this in Base too if it's going to be such an official interpretation; and maybe that's how we get nicer custom log level printing too (I.e. `Info+1` instead of `LogLevel(1)`)

Reply:
this is actually kinda unfortunate positioning.

because we want `@debugv 0` to be the same as `@debug` and to be the "most significant debug level"
Ideally we would have that `Debug +0` is in the middle of the range that highlights as `Debug`  so that `Debug+1` and `Debug-1` can both be considered as more or less significant debug-levelled values.

Reply:
I looked briefly at changing this, but it didn't seem like an easy/obvious fix with my first attempt, mainly because of the _directionality_ of log levels and filtering. Filtering always relies on filtering out _lower_ levels, so it doesn't work to just invert our verbosity to be + instead of -.

Reply:
Yeah, that is why i wish `Debug +0` was in the middle of the range that highlights as `Debug`.
Can we have this changed up-stream in Julia, and then just live with the bad highlighting until 1.9 comes out?

Reply:
see: https://github.com/JuliaLang/julia/pull/45708

Reply:
Thinking about this some more, I think I'd like to go back to the approach of using the `group` variable for verbosity. I think trying to leverage log levels has just muddied the waters too much. I also realized that there's one feature that "doesn't work" with the log-level approach: if I have a scenario like [here](https://github.com/JuliaWeb/HTTP.jl/blob/ec17649a49c078de5d173837080f38701d137c1c/src/Servers.jl#L393), where, in the normal case, I just want to ignore any errors, but if I'm in a "verbose" mode, I'd like to at least log those kinds of errors, it doesn't work w/ our current log-level approach. Because `@errorv 2 msg` just becomes `LogLevel.Error - 2`, and that means it will _always_ get printed, because it's "above" the default `Info` level.

So I think this has clarified in my mind that *verbosity* is really a 2nd dimension, in addition to log level, that I want to be able to filter on separately. i.e. if my `verbosity=1`, then I don't want to see the overly verbose `@errorv 2 msg` log. It also cements in my mind that log levels, while _kind of_ a proxy for verbosity, just aren't. They're categorical values that give specific "types" to log messages.

If no one is strongly opposed, I'm going to move the `LoggingExtras.@[x]v` macros back to that approach.

I think another benefit of that approach is we should "fix" the log level printing color issue? Since we're not modifying the actual log level at all?